### PR TITLE
support concurrency on rooted repositories 

### PR DESCRIPTION
**Pull request by smola**

### CODE REVIEW

**Comment by smola** [comment]
We have to add the lock service to `src-d/core-retrieval` before merging this.

File: archiver.go [hunk 1]
```diff
@@ -182,23 +190,31 @@ func (a *Archiver) pushChangesToRootedRepositories(j *Job, r *model.Repository,
...
+			log.Warn("failed to acquire lock", "root", ic.String(), "error", err)
+			continue
+		}
+
 		if err := a.pushChangesToRootedRepository(r, tr, ic, cs); err != nil {
 			err = ErrPushToRootedRepository.Wrap(err, ic.String())
 			a.notifyWarn(j, err)
 			failedInits = append(failedInits, ic)
-			//TODO: release lock
+			lock.Unlock()
```
Comment (ajnavarro) [inline archiver.go]: we should handle error here.

File: archiver.go [hunk 2]
```diff
@@ -182,23 +190,31 @@ func (a *Archiver) pushChangesToRootedRepositories(j *Job, r *model.Repository,
...
+
 		r.References = updateRepositoryReferences(r.References, cs, ic)
 		if err := a.dbUpdateRepository(r, now); err != nil {
 			err = ErrPushToRootedRepository.Wrap(err, ic.String())
 			a.notifyWarn(j, err)
 			failedInits = append(failedInits, ic)
 		}
-		//TODO: release lock
+
+		lock.Unlock()
```
Comment (ajnavarro) [inline archiver.go]: the same as previous comment.

File: archiver_test.go [hunk 3]
```diff
@@ -140,6 +141,42 @@ func (s *ArchiverSuite) TestFixtures() {
 	}
 }
 
+func (s *ArchiverSuite) TestBad() {
```
Comment (ajnavarro) [inline archiver_test.go]: What means TestBad? which is bad?

File: cli/borges/consumer.go [hunk 4]
```diff
@@ -31,6 +32,7 @@ func (c *consumerCmd) Execute(args []string) error {
 		core.ModelRepositoryStore(),
 		core.RootedTransactioner(),
 		borges.NewTemporaryCloner(core.TemporaryFilesystem()),
+		lock.NewLocal(),
```
Comment (ajnavarro) [inline cli/borges/consumer.go]: we should call here to the container to obtain a lock instance depending of the environment.

File: glide.lock [hunk 5]
```diff
@@ -108,10 +154,12 @@ imports:
   - repository
   - test
 - name: gopkg.in/src-d/framework.v0
-  version: 06c61f8f6a8913f3d4d465bab758ff06c0b3d920
+  version: 9bf13f3e57f5175ba43bc8c4def79a024041a57f
+  repo: https://github.com/smola/framework.git
```
Comment (ajnavarro) [inline glide.lock]: leftover in this PR? we should merge first framework and then upload this lock file.

File: glide.yaml [hunk 6]
```diff
@@ -32,10 +32,13 @@ import:
   - model
   - repository
 - package: gopkg.in/src-d/framework.v0
-  version: 06c61f8f6a8913f3d4d465bab758ff06c0b3d920
+  repo: https://github.com/smola/framework.git
```
Comment (ajnavarro) [inline glide.yaml]: See previous comment

File: archiver.go (hunk 1 above)
Comment (smola) [inline archiver.go]: In the current version, `Unlock()` never returns an error.

File: archiver.go (hunk 2 above)
Comment (smola) [inline archiver.go]: Same here.

File: archiver_test.go (hunk 3 above)
Comment (smola) [inline archiver_test.go]: Ouch! That was not supposed to be in the PR. It's just a local test I use to test repositories that fail.

File: glide.lock (hunk 5 above)
Comment (smola) [inline glide.lock]: Yes, changes to `src-d/framework` and `src-d/core-retrieval` are pending.

**Pull request merged**